core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -101,6 +101,53 @@ public static Snapshot oldestAncestor(Table table) {
     return ancestorsOf(start, lookup);
   }
 
+  /**
+   * Traverses the history of the table's current snapshot
+   * and finds the oldest ancestor snapshot after or equal to the timestamp in milliseconds.
+   *
+   * @param table           a table
+   * @param timestampMillis a timestamp in milliseconds
+   * @return the first snapshot from the given timestamp, or null if the current snapshot is older than the timestamp
+   * @throws IllegalStateException if the first snapshot from the given time can't be determined
+   */
+  public static Snapshot oldestAncestorFrom(Table table, Long timestampMillis) {
+    Snapshot currentSnapshot = table.currentSnapshot();
+    long timestamp = timestampMillis == null ? -1L : timestampMillis;
+    // If no snapshot exists or timestamp is higher than the current snapshot
+    if (currentSnapshot == null || currentSnapshot.timestampMillis() < timestamp) {
+      return null;
+    }
+
+    // Traverse to the oldest snapshot after or equal to the timestamp
+    Snapshot lastSnapshot = null;
+    for (Snapshot snapshot : currentAncestors(table)) {
+      if (snapshot.timestampMillis() == timestamp) {
+        return snapshot;
+      } else if (snapshot.timestampMillis() < timestamp) {
+        return lastSnapshot;
+      }
+      lastSnapshot = snapshot;
+    }
+
+    // Oldest snapshot of the table which is not expired
+    if (lastSnapshot != null) {
+      return lastSnapshot;
+    } else {
+      return oldestAncestor(table);
```

Comment:
How do we reach this branch? Shouldn't lastSnapshot have a value unless currentAncestors() is empty? And if currentAncestors is empty shouldn't we be returning null?

Reply:
Yes, I am reverting the fallback behaviour.